craft_parts/sources/base.py

```diff
@@ -0,0 +1,217 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Base classes for source type handling."""
+
+import abc
+import os
+import shutil
+from pathlib import Path
+from typing import List, Optional, Union
+
+import requests
+
+from craft_parts import utils
+from craft_parts.dirs import ProjectDirs
+from craft_parts.utils import url_utils
+
+from . import errors
+from .cache import FileCache
+from .checksum import verify_checksum
+
+
+class SourceHandler(abc.ABC):
+    """The base class for source type handlers.
+
+    Methods :meth:`check_if_outdated` and :meth:`update_source` can be
+    overridden by subclasses to implement verification and update of
+    source files.
+    """
+
+    # pylint: disable=too-many-arguments
+
+    def __init__(
+        self,
+        source: Union[str, Path],
+        source_dir: Union[str, Path],
+        *,
+        application_name: str = None,
+        source_tag: str = None,
+        source_commit: str = None,
+        source_branch: str = None,
+        source_depth: Optional[int] = None,
+        source_checksum: str = None,
+        command: str = None,
+        dirs: ProjectDirs = None,
+    ):
+        if not application_name:
+            application_name = utils.package_name()
+
+        if not dirs:
+            dirs = ProjectDirs()
+
+        self.source = str(source)
+        self.source_dir = str(source_dir)
+        self.source_tag = source_tag
+        self.source_commit = source_commit
+        self.source_branch = source_branch
+        self.source_depth = source_depth
+        self.source_checksum = source_checksum
+        self.source_details = None
+
+        self.command = command
+
+        self._application_name = application_name
+        self._dirs = dirs
+        self._checked = False
+
+    # pylint: enable=too-many-arguments
+
+    @abc.abstractmethod
+    def pull(self) -> None:
+        """Retrieve the source file."""
+
+    def check_if_outdated(
+        self, target: str, *, ignore_files: Optional[List[str]] = None
+    ) -> bool:
+        """Check if pulled sources have changed since target was created.
+
+        :param target: Path to target file.
+        :param ignore_files: Files excluded from verification.
+
+        :return: Whether the sources are outdated.
+
+        :raise errors.SourceUpdateUnsupported: If the source handler can't check if
+            files are outdated.
+        """
+        raise errors.SourceUpdateUnsupported(self.__class__.__name__)
+
+    def update(self):
+        """Update pulled source.
+
+        :raise errors.SourceUpdateUnsupported: If the source can't update its files.
+        """
+        raise errors.SourceUpdateUnsupported(self.__class__.__name__)
+
+
+class FileSourceHandler(SourceHandler):
+    """Base class for file source types."""
+
+    # pylint: disable=too-many-arguments
+    def __init__(
+        self,
+        source: Union[str, Path],
+        source_dir: Union[str, Path],
```

Comment:
```suggestion
        part_src_dir: Union[str, Path],
```

Reply:
`source_dir` is indeed a bad name, it looks like it's a part property when it's actually the part source directory. Will rename as suggested.